src/Substance.hs

```diff
@@ -4,535 +4,450 @@
+                      errls         = map (\(err1,t1) -> err1) errAndTypesLs
+                      err           = foldl (\err1 err2 -> err1 ++ err2) "" errls
+                      argTypes      = map (\(err1,t1) -> t1) errAndTypesLs
+                  in if (foldl (\b at1 -> b && isJust at1) True argTypes)
+                     then let argTypes2 = map (\a -> KT (fromJust a)) argTypes
+                              tls2      = map (\a -> KT a) tls
+                              sigma     = subst varEnv M.empty argTypes2 tls2
+                           in if (sigma == M.empty)
+                              then (err, Just t) -- err should be empty str
+                              else (err, Just (applySubst sigma t)) -- err should be empty str
+                     else (err, Nothing)
+
+applySubst :: M.Map Y Arg -> T -> T
+applySubst sigma (TTypeVar vt) = 
+           case sigma M.! (TypeVarY vt) of
+           AVar v -> error "Type var being mapped to variable in subst sigma, error in the TypeChecker!"
+           AT t   -> t
+applySubst sigma (TConstr (ConstructorInvoker t args pos)) =
+           let argsSub = map (applySubstHelper sigma) args
+           in TConstr (ConstructorInvoker t argsSub pos)
+
+applySubstHelper :: M.Map Y Arg -> Arg -> Arg
+applySubstHelper sigma (AVar v) = case sigma M.! (VarY v) of
+                                  res@(AVar v2) -> res
+                                  AT t -> error "Var being mapped to a type in subst sigma, error in the TypeChecker!"
+applySubstHelper sigma (AT t) = AT (applySubst sigma t)
+
+subst :: VarEnv -> M.Map Y Arg -> [K] -> [K] -> M.Map Y Arg
+subst varEnv sigma argTypes formalTypes = let types = zip argTypes formalTypes
+                                              sigma2 = foldl (substHelper varEnv) sigma types
+                                          in sigma2
+
+substHelper :: VarEnv -> M.Map Y Arg -> (K, K) -> M.Map Y Arg
+substHelper varEnv sigma ((Ktype aT), (Ktype fT)) = sigma
+substHelper varEnv sigma ((Ktype aT), (KT fT)) =
+                   error ("Argument type " ++ (show aT) ++ " doesn't match expected type " ++ (show fT))
+substHelper varEnv sigma ((KT aT), (Ktype fT)) =
+                   error ("Argument type " ++ (show aT) ++ " doesn't match expected type " ++ (show fT))
+substHelper varEnv sigma ((KT (TTypeVar atv)), (KT (TTypeVar ftv))) =
+                   substInsert sigma (TypeVarY ftv) (AT (TTypeVar atv))
+substHelper varEnv sigma ((KT (TTypeVar atv)), (KT (TConstr (ConstructorInvoker ftc argsFT pos)))) =
+                   error ("Argument type " ++ (show atv) ++ " doesn't match expected type " ++ (show ftc))
+substHelper varEnv sigma ((KT (TConstr (ConstructorInvoker atc argsAT pos))), (KT (TTypeVar ftv))) =
+                   substInsert sigma (TypeVarY ftv) (AT (TConstr (ConstructorInvoker atc argsAT pos)))
+substHelper varEnv sigma ((KT (TConstr (ConstructorInvoker atc argsAT pos1))), (KT (TConstr (ConstructorInvoker ftc argsFT pos2)))) =
```

Comment:
this case could be refactored:
* put the simple/successful if/then cases at the top
* i'm confused about the logic here. are atc and ftc allowed to be type variables? what if one of them is a variable and the other is a type? either way, please write out the logic here first, since it could get tricky